### PR TITLE
[#989]:Returns object for getContent() method

### PR DESCRIPTION
**Pull request by jimma**

### CODE REVIEW

**Review by spericas — CHANGES_REQUESTED** [review]
See `getContent` methods.

**Comment by jansupol** [comment]
The issue with these `getType` and `getGenericType` methods is that `EntityPart` used on the "incoming" side does not have the type and generic type information available after receiving the wired data. Perhaps a better solution would be to use GenericEntity as content? I assume that these methods are a reaction for `EntityPart.Builder content(T content, GenericType<T> type)`, perhaps the implementation can compose the `GenericEntity` from `content` and `type` if `content` is not a `GenericEntity` already?

**Comment by spericas** [comment]
Closing after not getting the necessary approvals.

**Comment by jimma** [comment]
> The issue with these `getType` and `getGenericType` methods is that `EntityPart` used on the "incoming" side does not have the type and generic type information available after receiving the wired data. Perhaps a better solution would be to use GenericEntity as content? I assume that these methods are a reaction for `EntityPart.Builder content(T content, GenericType<T> type)`, perhaps the implementation can compose the `GenericEntity` from `content` and `type` if `content` is not a `GenericEntity` yet?

@jansupol  Thanks for your input. My proposal is when we add some simple content like : https://github.com/eclipse-ee4j/jaxrs-api/issues/989#issuecomment-876247852, we need  a simple 'getContent' method to read it. When we set a string content, it has to wrap the string content to a InputStream result for 'getConent()' now. Is it better it simply returns a 'Object' ?

**Comment by jansupol** [comment]
@jimma This pull request introduces two independent things:
1) `getType()` & `getEntityType()` - see my comment
2) `Object getContent()`  - You can use [getContent(String.class)](https://github.com/eclipse-ee4j/jaxrs-api/blob/eb2891d33f8dfbde575b7f87e9abe823c7fe4be8/jaxrs-api/src/main/java/jakarta/ws/rs/core/EntityPart.java#L156), can you not?

**Comment by jimma** [comment]
@jansupol  Thanks for the input again.

> Now the discussion seems to be about a better return type or better convenient method. In your example, you use a client filter that indeed has the object. But the main usage of the method is IMO on the receiving side, where you do not have the deserialized object, you have the InputStream data.  There, the javadoc describes the behavior. Either you get the InputStream, or you need to use the message body provider to deserialize the input stream data to another java type.

 Having a better return type like ```object```, this will work for both sides. If I understand correctly, the EntityPart is for sending and receiving side, it would be good if we define a general return type instead of only for receiving side.

**Comment by jimma** [comment]
@jansupol  I changed this PR to add the `getContentStream()` and return `Object` for `getContent`.  Please review. Thanks.

File: jaxrs-api/src/main/java/jakarta/ws/rs/core/EntityPart.java
Comment (andymc12) [inline jaxrs-api/src/main/java/jakarta/ws/rs/core/EntityPart.java:183]:
```suggestion
     * an {@code IllegalStateException} if it is called after the {@code InputStream} has been
     * fully consumed already and has not been buffered prior to consuming.
```

File: jaxrs-api/src/main/java/jakarta/ws/rs/core/EntityPart.java
Comment (andymc12) [inline jaxrs-api/src/main/java/jakarta/ws/rs/core/EntityPart.java:144]:
```suggestion
     * an {@code IllegalStateException} if it is called after the {@code InputStream} has
     * been fully consumed already and has not been buffered prior to consuming.
```

**Comment by jimma** [comment]
@jansupol  @spericas @andymc12  Can you please review this again ?

File: jaxrs-api/src/main/java/jakarta/ws/rs/core/EntityPart.java
Comment (andymc12) [inline jaxrs-api/src/main/java/jakarta/ws/rs/core/EntityPart.java:112]: The text here seems too vague to me, and I worry that it might confuse users.  IIUC, the intent of the `getContent` method is:
(1) on the _sending_ side, this method returns the actual object (`String`, `File`, `Widget`, `byte[]`, etc.) that was passed in to the EntityPart's builder.
(2) on the _receiving_ side, this method is basically the same as `getContentStream`, but requires the user to cast it.

I think this method really only has any value on the sending side.  Maybe we could change this to just throw an `IllegalStateException` on the receiving side?

Alternatively, we could view the EntityPart as a "holder" for the entity object on both the sending and receiving sides - then it should always return the entity object (`String`, `Widget`, etc.) but only after a (not yet written) `bufferContent()` is invoked followed by `getContent(Class)` - similar to `Response#bufferEntity()` and `Response#readEntity(...)`.

Comment (jimma) [reply jaxrs-api/src/main/java/jakarta/ws/rs/core/EntityPart.java:112]: Thanks for review @andymc12 
> I think this method really only has any value on the sending side. Maybe we could change this to just throw an IllegalStateException on the receiving side?

Do you mean this method is only for sending side and calling this method throws IllegalStateException on the receiving side ? 

> Alternatively, we could view the EntityPart as a "holder" for the entity object on both the sending and receiving sides - then it should always return the entity object (String, Widget, etc.) but only after a (not yet written) bufferContent() is invoked followed by getContent(Class) - similar to Response#bufferEntity() and Response#readEntity(...).

+1 . I think this is a good idea.  Should we still keep 'getContentStream()' ?

**Comment by spericas** [comment]
@jimma @andymc12 @jansupol Given where we are in the release process, I'd suggest we revisit this post 3.1.

**Comment by andymc12** [comment]
@spericas I'm fine with moving this to 4.0.  Thanks.

File: jaxrs-api/src/main/java/jakarta/ws/rs/core/EntityPart.java
Comment (mkarg) [inline jaxrs-api/src/main/java/jakarta/ws/rs/core/EntityPart.java:114]: Typo "the an"

**Review by mkarg — CHANGES_REQUESTED** [review]

**Comment by mkarg** [comment]
@jakartaee/ee4j-jaxrs-committers How to proceed?

**Comment by spericas** [comment]
Closing given that it was never approved

**Pull request closed without merging**
